src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,209 @@
+
+namespace System.Net.Http
+{
+    internal partial class CurlHandler : HttpMessageHandler
+    {
+        [Conditional(VerboseDebuggingConditional)]
+        private static void IgnorePeerCert(EasyRequest easy)
+        {
+            easy.SetCurlOption(CURLoption.CURLOPT_VERIFYPEER, 0L);
+        }
+
+        /// <summary>
+        ///   Uses libssl's client certificate callback to provide the client certificate.
+        ///
+        ///   Works only for libcurl with an ssl-backend that is compatible with libssl
+        /// </summary>
+        private sealed class ClientCertProvider : IDisposable
+        {
+            private static readonly Interop.libcurl.ssl_ctx_callback s_sslCallback = CurlSslCallback;
+            private readonly GCHandle _gcHandle;
+            private SafeEvpPKeyHandle _privateKeyHandle = null;
+            private SafeX509Handle _certHandle = null;
+
+            private ClientCertProvider ()
+            {
+                _gcHandle = GCHandle.Alloc(this);
+            }
+
+            public void SetClientCertificateOptions(EasyRequest easy)
+            {
+                easy.SetCurlOption(CURLoption.CURLOPT_SSL_CTX_FUNCTION, s_sslCallback);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSL_CTX_DATA, GCHandle.ToIntPtr(_gcHandle));
+                IgnorePeerCert(easy);
+                VerboseTrace("setting client certificate options");
+            }
+
+            public void Dispose()
+            {
+                _gcHandle.Free();
+                if (_privateKeyHandle != null)
+                {
+                    _privateKeyHandle.Dispose();
+                }
+
+                if (_certHandle != null)
+                {
+                    _certHandle.Dispose();
+                }
+            }
+
+            private int TlsClientCertCallback(IntPtr ssl, out IntPtr certHandle, out IntPtr privateKeyHandle)
+            {
+                using (Interop.libssl.SafeSslHandle sslHandle = new Interop.libssl.SafeSslHandle(ssl))
+                {
+                    certHandle = IntPtr.Zero;
+                    privateKeyHandle = IntPtr.Zero;
+                    VerboseTrace("libssl's client certificate callback");
+                    ISet<string> issuerNames = GetRequestCertificateAuthorities(sslHandle);
+                    X509Certificate2 certificate;
+                    X509Chain chain;
+                    if (!GetClientCertificate(issuerNames, out certificate, out chain))
+                    {
+                        VerboseTrace("no cert or chain");
+                        return 0;
+                    }
+
+                    Debug.Assert(chain != null && chain.ChainElements.Count > 0, " a X509Chain received must not be empty");
+
+                    using (RSAOpenSsl rsa = certificate.GetRSAPrivateKey() as RSAOpenSsl)
+                    using (ECDsaOpenSsl ecdsa = certificate.GetECDsaPrivateKey() as ECDsaOpenSsl)
```

Comment:
One (or both) of them is guaranteed to return null.  To me this looks a little less bad than

```
using (RSAOpenSsl rsa = certificate.GetRSAPrivateKey() as RSAOpenSsl)
{
    if (rsa != null)
    {
        _privateKeyHandle = rsa.DuplicateKeyHandle();
    }
}

if (_privateKeyHandle == null)
{
    using (ECDsaOpenSsl ...)
    {
        ...
    }
}
```

And since there's no `IOpenSslAsymmetricAlgorithm`, or what have you, there's no polymorphic access to the DuplicateKeyHandle methods.